pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,326 @@
+			if err != nil {
+				errList = append(errList, fmt.Sprintf("parseAndValidateLocalProfile: %s", err))
+				continue
+			}
+			return localProfile, nil
+		}
+	}
+	return nil, fmt.Errorf("getLocalProfileConfig: all attempts failed: %s", strings.Join(errList, ";"))
+}
+
+func writeReceivedProtoMessageLocalProfile(localProfile *profile.LocalProfile) {
+	contents, err := proto.Marshal(localProfile)
+	if err != nil {
+		log.Errorf("writeReceivedProtoMessageLocalProfile Marshalling failed: %s", err)
+		return
+	}
+	writeProtoMessage(savedLocalProfileFile, contents)
+	return
+}
+
+//parseProfile process local and global profile configuration
+//must be called before processing of app instances from config
+func parseProfile(ctx *getconfigContext, config *zconfig.EdgeDevConfig) {
+	log.Functionf("parseProfile start: globalProfile: %s localProfile: %s",
+		ctx.globalProfile, ctx.localProfile)
+	if ctx.globalProfile != config.GlobalProfile {
+		log.Noticef("parseProfile: GlobalProfile changed from %s to %s",
+			ctx.globalProfile, config.GlobalProfile)
+		ctx.globalProfile = config.GlobalProfile
+	}
+	if ctx.localProfileServer != config.LocalProfileServer {
+		// read saved config on the first call
+		if !ctx.callProcessProfile {
+			if localProfile, err := readSavedLocalProfile(ctx, false); err != nil {
+				log.Functionf("parseProfile readSavedLocalProfile: %s", err)
+			} else if localProfile != nil {
+				log.Noticef("starting with localProfile %s", localProfile.LocalProfile)
+				ctx.localProfile = localProfile.LocalProfile
+			}
+			ctx.callProcessProfile = true
+		} else {
+			// if localProfileServer changed not in the first call of function
+			// clean saved proto message and set localProfile to empty
+			cleanSavedProtoMessage(savedLocalProfileFile)
+			ctx.localProfile = ""
+		}
+		log.Noticef("parseProfile: LocalProfileServer changed from %s to %s",
+			ctx.localProfileServer, config.LocalProfileServer)
+		ctx.localProfileServer = config.LocalProfileServer
+		if ctx.localProfileServer != "" {
+			// notify to reload local profile on profileServerChange
+			select {
+			case ctx.localProfileTrigger <- Notify{}:
+			default:
+			}
+		}
+	}
+	ctx.profileServerToken = config.ProfileServerToken
+	currentProfile := determineCurrentProfile(ctx)
+	if ctx.currentProfile != currentProfile {
+		// if we do not use LocalProfileServer set profile to GlobalProfile
+		log.Noticef("parseProfile: profile changed from %s to %s",
+			ctx.currentProfile, currentProfile)
+		ctx.currentProfile = currentProfile
+		publishZedAgentStatus(ctx)
+	}
+	log.Functionf("parseProfile done globalProfile: %s currentProfile: %s",
+		ctx.globalProfile, ctx.currentProfile)
+}
+
+//determineCurrentProfile return current profile based on localProfile, globalProfile and localProfileServer
+func determineCurrentProfile(ctx *getconfigContext) string {
+	if ctx.localProfileServer == "" || ctx.localProfile == "" {
```

Comment:
localProfileServer shouldn't be taken into account. Should it be cleared the state machine needs to take care of clearing localProfile.